### DISCUSSION TITLE
Proposal: Unified Runtime and Compile-Time Metaprogramming

### DISCUSSION THREAD

**Discussion by Blealtan** [dotnet/csharplang#3480] [Language Ideas]
> This proposal is based on a discussion at #3104. As suggested by @svick, I'm firing a new issue for my proposal on metaprogramming.

## Background

Currently, in C# we have a runtime metaprogramming mechanism that is Expression Tree, and a compile-time metaprogramming mechanism in preview status that is Source Generator. To me, they seem to have these disadvantages:

1. Expression Trees are old and not up-to-date. According to [this issue](https://github.com/dotnet/runtime/issues/27790) it is now archived and will not accept updates.
1. Expression Trees make it hard to wrap an expression into another. For example, if one wants to generate an Expression Tree containing a loop that executes a user-defined LambdaExpression many times, he/she will have to manually wrap the Expression object by calling the APIs directly.
1. Source Generator walks over the entire source tree, while in many cases a library developer may only want to touch a small fraction of them.
1. Source Generator works on Roslyn Syntax Trees and generates code in text form. This makes it hard to write since there will be no IntelliSense support IDE when the developer is writing generated code in a source generator.
1. Both being metaprogramming features, they are designed separately and differ a lot.

## Brief Introduction to what Scala 3.0 is doing

There is a new solution to metaprogramming proposed by the Scala community, which is now implemented as the metaprogramming feature set in Dotty, the experimental Scala 3.0 compiler. They published a paper on it that is [A Practical Unification of Multi-stage Programming and Macros](https://biboudis.github.io/papers/pcp-gpce18.pdf), which is worth reading.

For those who don't bother to read the full paper, I'm here introducing it in short. It is built on two operators: quoting (`'{ ... }` in both latest Dotty and the paper) and splicing (`${ ... }` in latest Dotty and `~` in the paper). Quoting turns what's inside into an `Expr[T]` object representing a typed expression, and splicing does the reverse. Phase Consistency Principle (PCP) is introduced to force any symbol to be declared and used in the same phase; phase level increments by one when it enters a quote or leaves a splice, and decrements by one vice versa. Below is an example of using it as a runtime metaprogramming feature, a.k.a. multi-stage programming.

![image](https://user-images.githubusercontent.com/6376623/82553684-18ffb680-9b97-11ea-865f-06a50797f873.png)

With another feature called `inline`, which is replacing a variable with the original side-effect-free expression provided in the function call and replace the function calls with function body in call-site. Dotty is able to provide macro functionality in a unified form. The key to compile-time metaprogramming with this notation is to allow top-level splicing: the compiler is exactly one lower phase level than the source code, thus code in top-level splicing is executed by the compiler, generates an expression and replaces the splicing in source code with it. Below is an example in the paper that does runtime and compile-time metaprogramming with the same implementation.

![image](https://user-images.githubusercontent.com/6376623/82557847-0d17f280-9b9f-11ea-89ec-aa7fd8a91e59.png)

There are many other aspects of this design, such as phase consistency of types, lifting and unlifting, etc. For these contents, please refer to the original paper or [Dotty Documentation](https://dotty.epfl.ch/docs/reference/metaprogramming/toc.html) for more details.

## Pros and Cons of this Metaprogramming Design as to C#

Here I share some personal considerations on this design:

* Pro: the same metaprogramming functionality would be available at both compile-time and runtime once implemented.
* Pro: both compile-time and runtime code generation are typed and safe.
* Pro: users are not required to produce AST-like structure by hand (quasi-quote out-of-the-box).
* Con: it'll be hard and unnatural to produce top-level structures, like namespaces, classes, or structs; in other words, it's only good at manipulating expressions and statements.

I might have missed many other points; please comment if you have more insights, especially the disadvantage side.

## Precedenting the Expression Trees and some of the Source Generators

If we adapt this metaprogramming design as-is, uses of Expression Trees will be possible to be replaced by using the quoting operator directly. Also, those libraries that are not meant to generate full class constructs will possibly benefit more from this metaprogramming design compared to the current Source Generators.

> Some personal mimic: I don't really like the `inline` design. It's ugly and looks not natural. But I can't come up with another design replacing it to enable macros/compile-time metaprogramming :(

**Comment by HaloFour** [comment]
One major advantage to source generators is that it is 100% outside of the language and requires no new grammar or semantics to work.  A built-in approach to metaprogramming like this requires coming up with a "language-within-a-language" and all of the grammar and behavior to support it and adding all of that to the spec.  It seems like significantly more work and to me feels like diminishing returns, especially after the initial scratch is "itched" by source generators.  Although having it built-into the language like this does lower the bar quite a bit for simpler cases.

**Comment by Blealtan** [comment]
@HaloFour It's true that it will be a lot of work, but compared to source generators it can actually do more. It enables embedded DSL, makes it possible to replace calls to library interfaces with generated code which the source generator can't, makes the generative library interface works both at compile-time and runtime out-of-the-box, etc.
It doesn't work only for simpler cases. If accessing the underlying syntax tree of a typed expression is possible (just like TASTy Reflection in Dotty), we can do anything a source generator does in a splicing construct (except generating new classes).

**Comment by HaloFour** [comment]
> It enables embedded DSL

Honestly, I find this a good reason to not do this.  I find Scala's myriad of ways of defining DSLs to make the language terribly confusing to me as a developer and also to tooling like IDEs.

> makes it possible to replace calls to library interfaces with generated code 

The team has explicitly stated that they don't like the idea of metaprogramming redefining how existing source will behave.  This is why `original`/`replace` syntax was taken out of source generators and it's no longer possible to redefine what a member does.

**Comment by Blealtan** [comment]
> they don't like the idea of metaprogramming redefining how existing source will behave

This statement made me suddenly clear why I don't like `inline` in Dotty. What if no `inline` at all? For example, to do a generative power implementation, we do

```
Expr<double> StagedPower(Expr<double> x, int n) {
    if (n == 0) return ‘{ 1.0 };
    var half = StagedPower(x, n / 2);
    if (n % 2 == 0)
        return '{ $half * $half };
    else
        return '{ $half * $half * $x };
}
...
double x = ...;
var y = ${ StagedPower('x, 42) };
...
```

With this kind of restriction, users would know explicitly that a piece of code is to be generated where the splice is placed. Is this still confusing to you as a developer? I do have few engineering experiences as I'm more a researcher than a developer, thus I can't really think the same way as you do.

**Comment by bbarry** [comment]
@Blealtan 

That is going to be possible with source generators as they stand today:

```csharp
[InlineParam(nameof(n), 42)]
double StagedPower(double x, int n) {
    if (n == 0) return 1.0;
    var half = StagedPower(x, n / 2);
    if (n % 2 == 0)
        return  half * half;
    else
        return half * half * x;
}

double pow42(double x) {
    return StagedPower_42(x);
}
```
   
of course I leave writing the source generator to someone else...

**Comment by Blealtan** [comment]
@bbarry 

> of course I leave writing the source generator to someone else...

Yes, and the problem is the pain for someone writing that source generator. It will require recognizing `InlineParam` attribute, generate a suffixed method (and possibly deal with name conflicts), **doing Partial Evaluation** to implement the `StagePower_42` as a highly optimized one (instead of really recursing at runtime) which is REALLY hard... So it won't work. FYI, Partial evaluation essentially turns

```csharp
double StagedPower(double x, int n) {
    if (n == 0) return 1.0;
    var half = StagedPower(x, n / 2);
    if (n % 2 == 0)
        return  half * half;
    else
        return half * half * x;
}
```

into doing something semantically equivalent to executing

```
Expr<double> StagedPower(Expr<double> x, int n) {
    if (n == 0) return ‘{ 1.0 };
    var half = StagedPower(x, n / 2);
    if (n % 2 == 0)
        return '{ $half * $half };
    else
        return '{ $half * $half * $x };
}
```

with `n = 42` to get an optimized `x^42` (no recursion, just series of multiplication and temp vars) instead of  `x^n, where n = 42` (recursive calls and branches). The latter one (your example in case of no partial evaluation) doesn't take full advantage of code generation; of course doing partial evaluation with a source generator IS possible, but that would be an even longer story.

**Comment by ntovas** [comment]
What about something like the the razor syntax?

```cs
Expr<double> StagedPower(Expr<double> x, int n) {
    @if (n == 0) 
         <code>return 1.0;</code>

   @{
    var half = StagedPower(x, n / 2);
    if (n % 2 == 0)
        <code>return  half * half;</code>
    else
        <code>return  half * half * $x;</code>
   }
}
```

**Comment by Blealtan** [comment]
@ntovas I'm neutral to the syntax; actually it looks fairly well to me using XML-like syntax. but is the `@{}` really meaningful? In the case of executed at runtime, it will be just a normal control flow assembling an expression tree.
